src/libraries/System.Console/src/System/IO/KeyParser.cs

```diff
@@ -0,0 +1,409 @@
+        {
+            return false;
+        }
+
+        if (input[SequencePrefixLength + digitCount] is VtSequenceEndTag or '^' or '$' or '@') // it's a VT Sequence like ^[[11~ or rxvt like ^[[11^
+        {
+            int sequenceLength = SequencePrefixLength + digitCount + 1;
+            if (!TryMapUsingTerminfoDb(buffer.AsMemory(startIndex, sequenceLength), terminalFormatStrings, ref key, ref isShift, ref isAlt, ref isCtrl))
+            {
+                key = MapEscapeSequenceNumber(byte.Parse(input.Slice(SequencePrefixLength, digitCount)));
+
+                if (key == default)
+                {
+                    return false; // it was not a known sequence
+                }
+            }
+
+            if (input[SequencePrefixLength + digitCount] is '^' or '$' or '@') // rxvt modifiers
+            {
+                Apply(MapRxvtModifiers(input[SequencePrefixLength + digitCount]), ref isShift, ref isAlt, ref isCtrl);
+            }
+
+            startIndex += sequenceLength;
+            return true;
+        }
+
+        // If Sequence Number is not followed by the VT Seqence End Tag,
+        // it can be followed only by a Modifier Separator, Modifier (2-8) and Key ID or VT Sequence End Tag.
+        if (input[SequencePrefixLength + digitCount] is not ModifierSeparator
+            || SequencePrefixLength + digitCount + 2 >= input.Length
+            || !char.IsBetween(input[SequencePrefixLength + digitCount + 1], '2', '8')
+            || (!char.IsAsciiLetterUpper(input[SequencePrefixLength + digitCount + 2]) && input[SequencePrefixLength + digitCount + 2] is not VtSequenceEndTag))
+        {
+            return false;
+        }
+
+        ConsoleModifiers modifiers = MapXtermModifiers(input[SequencePrefixLength + digitCount + 1]);
+
+        key = input[SequencePrefixLength + digitCount + 2] is VtSequenceEndTag
+            ? MapEscapeSequenceNumber(byte.Parse(input.Slice(SequencePrefixLength, digitCount)))
+            : MapKeyIdOXterm(input[SequencePrefixLength + digitCount + 2], terminalFormatStrings.IsRxvtTerm);
+
+        if (key != default)
+        {
+            startIndex += SequencePrefixLength + digitCount + 3; // 3 stands for separator, modifier and end tag or id
+            Apply(modifiers, ref isShift, ref isAlt, ref isCtrl);
+            return true;
+        }
+
+        return false;
+
+        static bool TryMapUsingTerminfoDb(ReadOnlyMemory<char> inputSequence, ConsolePal.TerminalFormatStrings terminalFormatStrings,
+            ref ConsoleKey key, ref bool isShift, ref bool isAlt, ref bool isCtrl)
+        {
+            // Check if the string prefix matches.
+            if (terminalFormatStrings.KeyFormatToConsoleKey.TryGetValue(inputSequence, out ConsoleKeyInfo consoleKeyInfo))
+            {
+                key = consoleKeyInfo.Key;
+                Apply(consoleKeyInfo.Modifiers, ref isShift, ref isAlt, ref isCtrl);
+                return true;
```

Comment:
Why not return `consoleKeyInfo` instead of re-creating it on `StdInReader.ReadKey()`?

Reply:
Good idea, done!